### PR TITLE
[polygon] utils for geometries crossing the 180th meridian

### PR DESCRIPTION
**Pull request by Pessimistress**
`cutPolylineByMercatorWorld`:

math-wrap-mercator-01

`cutPolygonByMercatorWorld`:

math-wrap-mercator-03

### CODE REVIEW

**Comment by Pessimistress** [comment]
Note that this algorithm does not handle all polygon edge cases correctly, namely polygons that cross multiple worlds with holes:

applepeel

File: modules/polygon/src/cut-by-mercator-world.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    push(positions, p);
+    // Project the first vertex to the edge of the map
+    p[0] = firstLng;
+    push(positions, p);
+  }
+}
+
+function shiftLongitudesIntoRange(positions, size) {
+  let refLng;
+  const n = positions.length;
```
Comment (ibgreen) [inline modules/polygon/src/cut-by-mercator-world.js:137]: Please don't...  

I could buy `const positionCount = positions.length / size;`, that could make things more readable.

File: modules/polygon/src/cut-by-mercator-world.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    if ((refLng + 180) % 360 !== 0) {
+      break;
+    }
+  }
+
+  const delta = -Math.round(refLng / 360) * 360;
+  if (delta === 0) {
+    return;
+  }
+  for (let i = 0; i < n; i += size) {
```
Comment (ibgreen) [inline modules/polygon/src/cut-by-mercator-world.js:152]: Nit: I feel this is easier to read if we write

```
for (i = 0; i < positionCount; ++i) {
   positions[i * size + 0] += delta;
}
```

**Comment by Pessimistress** [comment]
@ibgreen More explanation regarding the poles:

Here is the raw plot of two GeoJSON polygons: 1) the Arctic circle 2) the Antarctic continent

image

In Web Mercator projection, each pole becomes a line. To tesselate such a polygons correctly, we need to insert the edge of map into the polygon. This requires adding two vertices that represent the pole, by drawing a perpendicular line to the Mercator map edge from a selected vertex on the ring.

image


`findSplitIndex`: We select the insertion position carefully so that the inserted line segments do not intersect with the ring itself. This is ensured by `findSplitIndex`, which returns the vertex closest to the pole.

`MAX_LATITUDE`: Latitude=90 degrees projects to infinity in Web Mercator, so when we insert the "pole", its latitude must be clamped. Theoretically any number smaller than 90 will do. `85.051129` is used by Google Maps/Mapbox because it yields a square world.

After being processed by `cutPolygonByMercatorWorld`:

image

File: modules/polygon/docs/api-reference/cut-polygon-by-mercator-world.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# cutPolygonByMercatorWorld
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/polygon/docs/api-reference/cut-polygon-by-mercator-world.md:1]: As discussed below this operation is Mercator specific. Renaming to `cutPolygonByMercatorBounds`.

**Pull request merged**